src/librustc_front/lowering.rs

```diff
@@ -1215,7 +1220,13 @@ pub fn lower_expr(lctx: &LoweringContext, e: &Expr) -> P<hir::Expr> {
                                 maybe_expr.as_ref().map(|x| lower_expr(lctx, x)))
             }
             ExprParen(ref ex) => {
-                return lower_expr(lctx, ex);
+                // merge attributes into the inner expression.
+                return lower_expr(lctx, ex).map(|mut ex| {
+                    ex.attrs.update(|attrs| {
+                        attrs.prepend(e.attrs.clone())
```

Comment:
hmm, it might be good to have a test exercising the ordering enforced by this code.

(I did see a test with a generic `#[attr]`, along the lines of `#[attr] ( #[attr] expr )`, but I didn't see anything that ensured that a prepend is occurring here rather than an append.)

Having said that, it isn't really that big a deal (especially since it may be non-trivial to actually make such a test... I guess one could do it by applying an `#[allow(foo)]` at the outer point followed by a `#[deny(foo)]` on the inner, and vice-versa, and making sure one gets the expected effect.